sklearn/ensemble/_bagging.py

```diff
@@ -126,7 +123,11 @@ def _parallel_build_estimators(
         )
 
         # Draw samples, using sample weights, and then fit
+        support_sample_weight = has_fit_parameter(
+            ensemble.base_estimator_, "sample_weight"
+        )
```

Comment:
This doesn't seem to support a meta-estimator which forwards sample-weight to the child estimator. Should we instead try to route metadata, and see if sample-weight gets routed?

cc @glemaitre @thomasjpfan

Reply:
As noted in https://github.com/scikit-learn/scikit-learn/pull/24250/files/04882927e1c98adb0e9b62657c9d491067cef8c3#r963820342 this is a bit more involved.  I'll continue the discussion in the other thread.

Reply:
I'll respond here to not overload the other thread. Here are the two situations I see:

### Case 1: `base_estimator_` is a meta estimator.

If `base_estimator_` requests for `sample_weights`, then pass it in. This is the same logic as we have for other meta-estimators.

### Case 2: `base_estimator_` is a regular estimator.

If `base_estimator` has a `sample_weight` fit parameter, bagging wants to pass in sample weights independent of the metadata request to simulate the bagging process. For backward compatibility, I think we need to preserve this behavior. Although, this behavior is inconsistent with metadata routing API. For example, on `main`, the following will pass in sample weights to `SVR` even if it is not requested:

```python
regr = BaggingRegressor(base_estimator=SVR())
```

Do we want to require `BaggingRegressor` to request sample weights to use the "bagging with sample weight" feature?

```python
regr = BaggingRegressor(base_estimator=SVR().fit_requests(sample_weight=True))
```

Reply:
Personally, I'm not in favor of making the routing of sample weights dependent on whether the base estimator is a meta estimator or not. Maybe there is precedence for that but at as a user, I would be surprised if the behavior changed in such a fashion just because I swapped one base estimator for another one.

Let's list all the factors we have:

1. Base estimator is a meta estimator or not
2. Base estimator supports sample weights (as determined by its signature)
3. Base estimator requested sample weight (as per fit request)
4. `bootstrap=True` or `False`
5. User passed sample weight to fit or not

As mentioned, factor 1 should not matter IMHO, leaving us to resolve the others.

Furthermore, we have 2 choices to make when it comes to using sample weights:

a. Sample via sample weights or via indexing
b. Pass sample weight to base estimator or not

I'm not really clear on what the point is of having sampling via sample weights if `bootstrap=False`. Is that not functionally equivalent to sampling via indexing, only potentially slower? Can we remove that branch completely? For my proposal, I assume that the two approaches are indeed equivalent.

With that out of the way, my proposal would be:

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">

<colgroup>
<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">BE supports SW</th>
<th scope="col" class="org-left">BE req SW</th>
<th scope="col" class="org-left">bootstrap</th>
<th scope="col" class="org-left">user passes SW</th>
<th scope="col" class="org-left">outcome</th>
</tr>
</thead>

<tbody>
<tr>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">pass SW to BE, sample via SW</td>
</tr>


<tr>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">pass SW to BE, sample via indexing</td>
</tr>


<tr>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">don&rsquo;t pass SW, sample via SW</td>
</tr>


<tr>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">don&rsquo;t pass SW, sample via SW</td>
</tr>


<tr>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">don&rsquo;t pass SW, sample via indexing</td>
</tr>


<tr>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">don&rsquo;t pass SW, sample via SW</td>
</tr>


<tr>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">T</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">F</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">T</td>
<td class="org-left">error</td>
</tr>


<tr>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">F</td>
<td class="org-left">sample via indexing</td>
</tr>
</tbody>
</table>

(BE = base estimator, SW = sample weight)

If I'm not mistaken, this preserves the current behavior but it opens the possibility for the user to control whether they want their sample weights to be passed to the base estimator or not.